src/wasmtime/component.rs

```diff
@@ -140,60 +193,61 @@ impl WasmValue for component::Val {
     ) -> Result<Self, WasmValueError> {
         ensure_type_kind(ty, WasmTypeKind::List)?;
-        ty.unwrap_list()
-            .new_val(vals.into_iter().collect())
-            .map_err(WasmValueError::other)
+        Ok(Self::List(vals.into_iter().collect()))
     }
     fn make_record<'a>(
         ty: &Self::Type,
         fields: impl IntoIterator<Item = (&'a str, Self)>,
     ) -> Result<Self, WasmValueError> {
-        ty.unwrap_record()
-            .new_val(fields)
-            .map_err(WasmValueError::other)
+        ensure_type_kind(ty, WasmTypeKind::Record)?;
+        let values: Vec<(String, Self)> = fields
+            .into_iter()
+            .map(|(name, val)| (name.to_string(), val))
+            .collect();
+
+        Ok(Self::Record(values))
     }
     fn make_tuple(
         ty: &Self::Type,
         vals: impl IntoIterator<Item = Self>,
     ) -> Result<Self, WasmValueError> {
-        ty.unwrap_tuple()
-            .new_val(vals.into_iter().collect())
-            .map_err(WasmValueError::other)
+        ensure_type_kind(ty, WasmTypeKind::Tuple)?;
+        Ok(Self::Tuple(vals.into_iter().collect()))
     }
     fn make_variant(
         ty: &Self::Type,
         case: &str,
         val: Option<Self>,
     ) -> Result<Self, WasmValueError> {
-        ty.unwrap_variant()
-            .new_val(case, val)
-            .map_err(WasmValueError::other)
+        ensure_type_kind(ty, WasmTypeKind::Variant)?;
+        Ok(Self::Variant(case.to_string(), val.map(Box::new)))
     }
     fn make_enum(ty: &Self::Type, case: &str) -> Result<Self, WasmValueError> {
-        ty.unwrap_enum()
-            .new_val(case)
-            .map_err(WasmValueError::other)
+        ensure_type_kind(ty, WasmTypeKind::Enum)?;
+        Ok(Self::Enum(case.to_string()))
     }
     fn make_option(ty: &Self::Type, val: Option<Self>) -> Result<Self, WasmValueError> {
-        ty.unwrap_option()
-            .new_val(val)
-            .map_err(WasmValueError::other)
+        ensure_type_kind(ty, WasmTypeKind::Option)?;
+        Ok(Self::Option(val.map(Box::new)))
     }
     fn make_result(
         ty: &Self::Type,
         val: Result<Option<Self>, Option<Self>>,
     ) -> Result<Self, WasmValueError> {
-        ty.unwrap_result()
-            .new_val(val)
-            .map_err(WasmValueError::other)
+        ensure_type_kind(ty, WasmTypeKind::Result)?;
+        Ok(match val {
+            Ok(val) => Self::Result(Ok(val.map(Box::new))),
+            Err(val) => Self::Result(Err(val.map(Box::new))),
+        })
     }
     fn make_flags<'a>(
         ty: &Self::Type,
         names: impl IntoIterator<Item = &'a str>,
     ) -> Result<Self, WasmValueError> {
-        ty.unwrap_flags()
-            .new_val(&names.into_iter().collect::<Vec<_>>())
-            .map_err(WasmValueError::other)
+        ensure_type_kind(ty, WasmTypeKind::Flags)?;
+        Ok(Self::Flags(
+            names.into_iter().map(|n| n.to_string()).collect(),
+        ))
```

Comment:
Here is an example of the type checking being relaxed: [`Flags::new_val`](https://docs.rs/wasmtime/19.0.2/wasmtime/component/types/struct.Flags.html#method.new_val) which was called at 195 would validate that `names` matched the defined flag names. The same applies to many of the `make_*` diffs above.

Reply:
Oh, I was totally on the wrong track about what you were referring to.

Thanks for pointing me at this. These kinds of changes/regressions I would want to correct before the PR lands.

I can work backwards from a stack trace to see when these are being called when the crate is tested and see how more tests would have caught the regression.

Reply:
Interesting. I see the merge of the make_list test. That's confirming the composite types match up.

Wonder if there is a way to make the inputs and outputs all strings, like the other tests. Or will these require coded calls to make_* for any valid and invalid combinations. Guess they can still be put into a table of inputs and expected type results.

Reply:
How about added checks along these lines?
```rust
    fn make_list(
        ty: &Self::Type,
        vals: impl IntoIterator<Item = Self>,
    ) -> Result<Self, WasmValueError> {
        ensure_type_kind(ty, WasmTypeKind::List)?;
        let component::Type::List(list) = ty else {
            // TBD this could be where the error is returned rather
            // than having to make the call to ensure_type_kind.
            panic!();
        };
        let elem_type = list.ty();
        vals.into_iter()
            .map(|val| {
                is_value_type(&val, &elem_type)?;
                Ok(val)
            })
            .collect::<Result<Vec<_>, WasmValueError>>()
            .map(Self::List)
    }
```
and
```rust
fn is_value_type(val: &component::Val, ty: &component::Type) -> Result<(), WasmValueError> {
    if val.kind() != ty.kind() {
        return Err(WasmValueError::wrong_value_type(ty, val));
    }
    match val {
        component::Val::List(vals) => {
            let component::Type::List(list) = ty else {
                unreachable!(); // kind to kind check should make this unreachable
            };
            let elem_type = list.ty();
            for val in vals {
                is_value_type(val, &elem_type)?;
            }
        },
        component::Val::Record(_vec_string_val) => todo!(),
        component::Val::Tuple(_vec_val_) => todo!(),
        component::Val::Variant(_string, _option_box_val__) => todo!(),
        component::Val::Enum(_string) => todo!(),
        component::Val::Option(_option_box_val__) => todo!(),
        component::Val::Result(_result_option_box_val_option_box_val) => todo!(),
        component::Val::Flags(_vec_string_) => todo!(),
        component::Val::Resource(_resourceany) => todo!(),

        // Any leaf variant type was already tested with the kind comparisons above.
        _ => {},
    }
    Ok(())
}
```
I don't like the almost duplication of code, but in the first case, I wanted to collect the values and in the second case I didn't. Open to ideas...

Reply:
I just had a thought, it could save a lot of code if the make_list creates the List but before returning it, checks that it ends up being a matching type. If the time spent and heap allocations made for the failure case is acceptable in this scenario.

Reply:
I haven't added any tests. I can understand why there weren't tests for this verify issue before because the verify was being done by the `wasmtime` library, not this one. Now that this is responsible for verifying variables are being created correctly, some more negative tests may be in order.